### PR TITLE
fix(focus-trap): add focus trap component and remove util function FE-3586

### PR DESCRIPTION
**Pull request by edleeks87**
BREAKING CHANGE: the focus trap util function has been removed

### Proposed behaviour

The `focusTrap` util does not support the correct tab order if a disabled component becomes active,
it has been removed and the functionality has been moved into an internal component that is composed
into the `Modal` based components

### Current behaviour

The `focusTrap` util exists there is no component

### Checklist

- [x] Commits follow our style guide
- [x] Screenshots are included in the PR if useful
- [x] All themes are supported if required
- [x] Unit tests added or updated if required
- [x] Cypress automation tests added or updated if required
- [x] Storybook added or updated if required
- [x] Typescript `d.ts` file added or updated if required
- [x] Carbon implementation and Design System documentation are congruent

### Testing instructions

The modal based stories will need to be checked to ensure they function as before, the `editable` `Dialog` story has been updated to show that it now supports tab ordering when disabled components are enabled

https://codesandbox.io/s/carbon-quickstart-forked-eeo0w?file=/src/index.js

### CODE REVIEW

**Comment by edleeks87** [comment]
> Good day, There appears to be bug in the Dialog -> editable MDX example.
> 
> Dialog_editable_bug

Good spot I'll make sure the radio buttons are wired up

**Review by DipperTheDan — DISMISSED** [review]

File: src/__internal__/focus-trap/focus-trap.component.js [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+  }, [firstElement, lastElement, focusableElements, bespokeTrap]);
+
+  return <div ref={ref}>{children}</div>;
+};
+
+FocusTrap.propTypes = {
+  children: PropTypes.node.isRequired,
+  autoFocus: PropTypes.bool,
+  focusFirstElement: PropTypes.func,
+  bespokeTrap: PropTypes.func,
```
Comment (mkrds) [inline src/__internal__/focus-trap/focus-trap.component.js:115]: I'd write prop descriptions here, as I'm not really sure what `bespokeTrap` does unless I check out the `FocusTrap` code

File: src/__internal__/focus-trap/focus-trap.component.js [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+  }, [children, hasNewInputs]);
+
+  useEffect(() => {
+    if (autoFocus && firstOpen.current && (focusFirstElement || firstElement)) {
+      setElementFocus(focusFirstElement || firstElement);
+      firstOpen.current = false;
+    }
+  }, [autoFocus, firstElement, focusFirstElement]);
+
+  useEffect(() => {
```
Comment (mkrds) [inline src/__internal__/focus-trap/focus-trap.component.js:63]: Is it necessary to run `ev.preventDefault()` and `setElementFocus` in cases when its not a last/first element?
I think when we're let's say on 3rd input out of 10 in `Dialog`, and press `Tab` then the default browser bevahiour would be just fine wouldnt it?

Comment (edleeks87) [reply src/__internal__/focus-trap/focus-trap.component.js:63]: It is necessary to run it unfortunately, as when a disabled element becomes enabled it is not included in the tab order as in the gif in the PR description. Even when you manually set a tabIndex of 0 when it becomes enabled it still doesn't include it until the Modal is closed and re-opened

File: src/__internal__/focus-trap/focus-trap.component.js [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    };
+  }, [firstElement, lastElement, focusableElements, bespokeTrap]);
+
+  return <div ref={ref}>{children}</div>;
+};
+
+FocusTrap.propTypes = {
+  children: PropTypes.node.isRequired,
+  autoFocus: PropTypes.bool,
+  focusFirstElement: PropTypes.func,
```
Comment (mkrds) [inline src/__internal__/focus-trap/focus-trap.component.js:114]: Is it always meant to be a function? I can't pass a ref to the element that I want to be focused first?

Comment (edleeks87) [reply src/__internal__/focus-trap/focus-trap.component.js:114]: Discussed privately, the components that use it currently use callback refs but I have expanded the prop type to support object refs as well

**Review by DipperTheDan — APPROVED** [review]

**Review by mkrds — APPROVED** [review]

**Review by DipperTheDan — DISMISSED** [review]

**Review by mkrds — DISMISSED** [review]

**Review by DipperTheDan — DISMISSED** [review]

**Review by Alec-Walker-Sage — DISMISSED** [review]

**Review by DipperTheDan — APPROVED** [review]

**Review by Alec-Walker-Sage — APPROVED** [review]

**Pull request merged**